### ISSUE TITLE
"Increase information density—with ASCII art!" is problematic for accessibility

### ISSUE THREAD

**Issue by dbjorge** [cli-guidelines/cli-guidelines#103]
There is an item in the [Output section](https://clig.dev/#output) that reads:

```markdown
**Increase information density—with ASCII art!** For example, `ls` shows permissions in a scannable way. When you first see it, you can ignore most of the information. Then, as you learn how it works, you pick out more patterns over time.
```

Unfortunately, while ASCII art can be great for improving information density for some users, it is generally problematic for accessibility and inclusivity:

* Screen reader users will generally have an experience ranging from "much more unpleasant than a visual user" (eg, with an ASCII art table) to "completely unusable" (eg, for most ASCII art progress indicators or product logos). Sections 5.3.2 and 5.4 of [this research paper](https://dl.acm.org/doi/fullHtml/10.1145/3411764.3445544) are good overviews of these issues.
* It tends to create/exacerbate formatting issues with text wrapping at low line lengths, which disproportionately impacts users that need to use high zoom levels and/or font sizes
* Very high information density can also make it harder for folks with certain types of neurodivergence to parse the output

I think it would be preferable to rewrite the ASCII art item along these lines:

```markdown
**Avoid using ASCII art!** This includes product logos in help and splash screens, progress indicators, and table-like formatting. ASCII art generally creates accessibility problems. For example, screen readers will announce ASCII art character by character instead of as a whole, and users that require high font sizes might see mangled output due to line wrapping.

**Table-like output requires extra caution.** Sometimes, a tool's output is naturally tabular, and it can make sense to support output formatted like a table. However, users of accessibility technologies (for example, screen readers) will not generally be able to navigate tabular CLI output column-wise or row-wise - they will be forced to read the output line-by-line. To make tabular output more inclusive:

* **Allow the user to control the order and visibility of columns.** For example, `git log` provides both a `--format` flag that enables exact control over what data is displayed and several shorthand flags for different levels of information density.
* **Do not use unicode formatting characters to separate columns visually.** Unnecessary characters like `╣` create noise for ATs.
* **Support exporting tabular output to a tabular data format.** Users of accessibility technologies will be better able to navigate your table row-wise or column-wise if they can export it to, for example, a `.csv` file or an `.html` file with an embedded table.
* **Any output visible in a table should also be present in other structured output formats.** For example, if you support a `--json` option, ensure that all the output exposed as a table is also exposed there.
* **Avoid using tables to format non-tabular data.** For example, do not copy the default no-argument behavior of `ls`.
```

---

I considered adding this as a comment on #5, but I thought this felt more immediately actionable when taken on its own, so I made it a separate issue.

**Comment by severak** [comment]
There is already section about disabling of color output. I think we can elaborate in similar style:

If the `USING_SCREENREADER` environment variable is set, output should be screenreader friendly.

For example, instead of using animated progress bar (`[=====> ]`) it would output something like:

```
10 % done
20 % done
30 % done
```

while it's doing own thing. Ideally - these screenreader alternatives to ASCII would be incorporated in libraries so they will be standardized to most usable variant. 

I think sometimes ASCII art can be valuable and actually makes sense in some cases (for example `tree` command) .

However I have exactly zero experience with using screenreader so I can't tell if this was done before.

**Comment by l0b0** [comment]
It's weird that the ASCII art example is tabular output - a table is not ASCII art, it's literally just aligned text.

**Comment by eqn-group** [comment]
I think if you are blind, this program is not for you. if this program is an issue use different tool. This must be left to the program author if they want ascii art or whatever.

**Comment by ssokolow** [comment]
> I think if you are blind, this program is not for you. if this program is an issue use different tool. This must be left to the program author if they want ascii art or whatever.

The point is that it shouldn't be encouraged in a list of general-purpose guidelines because it's not something that's virtuous in all situations.

**Comment by Garbee** [comment]
One reason using an environment variable, such as `USING_SCREENREADER`, to determine what to do does not work is; privacy. We can't expect that people who are using assistive technology expose that fact to every program running on their machine. As it could then be sucked up and leveraged to advertise to them or raise prices on them (legal or not.)

Software needs to be build in a privacy respectful way that also provides people what they want. Time and time again it has been proven that if you build accessible software by default then everyone using the software benefits. Simple software is better for everyone.

It is OK for a tool to have a specific flag that may enable this mode. (GitHub has one while experimenting with a more accessible interface.) But there shouldn't be some "global" since then it becomes too easy for other tools to maliciously check for its existence and use that information against someone.

**Comment by Garbee** [comment]
@ssokolow Regardless of anything, an *accessible and usable experience* for everyone should be _the default_. Is that something to actually debate? The less accessible forms should be opt-in rather than opt-out.

In other words, we should _enable features that degrade screen reader capability_ rather than have to opt out of it.

Is there really a problem with building by default for the widest possible audience?

**Comment by ssokolow** [comment]
True. I was just countering your reasoning for `USING_SCREENREADER`.

I would also argue that preferring use of ASCII art *as a general principle* runs counter to point 2 of the original statement of the UNIX philosophy:

> Expect the output of every program to become the input to another, as yet unknown, program. Don't clutter output with extraneous information. Avoid stringently columnar or binary input formats. Don't insist on interactive input.

However, one can also argue that, if these concerns really *are* in tension, then making things like progress bars opt-in is sort of like arguing that the majority of users must know that Windows 3.1 is available if you type `win` to protect users whose braille displays only work with MS-DOS.

The general assumption for a CLI interface set by POSIX's history is that, if a user comfort feature isn't on by default, then it doesn't exist and it's not worth digging through the manual looking for it.

To speak by analogy, suppose screen readers are tripped up by terminal escapes for colorizing output. The solution isn't to default to `TERM=vt100` for everyone and expect the users who don't need assistive technologies to RTFM and discover the iceberg of UX improvements available to them, but to either improve communication with the screen reader or to improve the ability to automate detecting when accessible mode is needed. (Similar to how programs already use `isatty(3)` to automatically switch their default behaviours when piped.)

**Comment by ssokolow** [comment]
> This is the most critical part of that response I feel. And it is where everything falls apart. I read this in context as, "If I don't get my progress bar and arrow key selection then it's worse for everyone."

As taken in the context of my first post from 2024, it's meant to say that a **general** prohibition is wrong. Naturally, it will depend on the context.

(And, in all honesty, though you meant "If I don't get my progress bar and arrow key selection then it's worse for everyone." to be hyperbolic, it's true for some contexts. Depending on the goal of the UI, that may be the non-negotiable part in some kind of workflow optimization, with the question being whether you get a TUI at all or an X11 GUI exported via `ssh -X`. Heck, I've worked on Qt GUIs where a core part of the human performance optimization is being able to sort images with a single press of one of the numpad keys, and you hit Backspace to Undo if your fingers get ahead of your brain.)

> But the default should not be adding junk and things that can't be understood by assistive technology in the way just because it looks good.

Again, if that principle had been followed in such a simplistic fashion in the past, we'd still be on monochrome `TERM=tty` without cursor motion beyond `\r` because there was no doubt a time when various terminal features were to complex for some assistive technology.

While more subtly so, your stance would be just as flawed as "Increase information density—with ASCII art!" in the context of a simple, blanket dictum.
